File: Binary_Tree_Level_Order_Traversal/binary_tree_level_order_traversal.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> levels = new LinkedList<>();
        if (root != null) {
            levels.offer(root);
        }
        while (levels.size() > 0) {
            List<Integer> current_level = new ArrayList<>();
            int level_size = levels.size();
            for (int i = 0; i < level_size; i++) {
                TreeNode current = levels.poll();
                current_level.add(current.val);
                if (current.left != null) {
                    levels.offer(current.left);
                }
                if (current.right != null) {
                    levels.offer(current.right);
                }
            }
            result.add(current_level);
        }
        return result;
    }
}

File: Binary_Tree_Level_Order_Traversal/binary_tree_level_order_traversal.cpp
#include <vector>
#include <queue>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        std::vector<std::vector<int>> result;
        std::queue<TreeNode*> levels;
        if (root) {
            levels.push(root);
        }

        while (!levels.empty()) {
            std::vector<int> current_level;
            int level_size = int(levels.size());
            for (int i = 0; i < level_size; i++) {
                TreeNode *current = levels.front();
                levels.pop();
                if (current->left) {
                    levels.push(current->left);
                }
                if (current->right) {
                    levels.push(current->right);
                }
                current_level.push_back(current->val);
            }
            result.push_back(current_level);
        }
        return result;
    }
};
